vidyut-lipi/src/lib.rs

```diff
@@ -1,118 +1,551 @@
+use std::error::Error;
+use std::fs::File;
+use std::io::BufReader;
+use std::path::Path;
+
+const IMPLICIT_VIRAMA: u8 = 0x01;
+const IMPLICIT_A: u8 = 0x02;
+const IMPLICIT_UNKNOWN: u8 = 0x07;
+const TOK_IS_VOWEL: u8 = 0x03;
+const TOK_IS_CONSONANT: u8 = 0x04;
+const TOK_IS_DIGIT: u8 = 0x05;
+const TOK_UNKNOWN_TYPE: u8 = 0x06;
+const TOK_IS_SIGN: u8 = 0x08;
+const TOK_IS_NUMERAL: u8 = 0x09;
+const TOK_TRANSPARENT_TYPE: u8 = 0x0A;
+const TOK_LEXICAL_BREAK: u8 = 0x0B;
+const TOK_ENCODING_ESCAPE: u8 = 0x0C;
+
+
+#[derive(Debug, Serialize, Deserialize)]
+struct Encoding {
+    enc: Vec<String>,
+    token: String,
+    ttype: String,
+    #[serde(default)]
+    assoc: String,
 }
 
-fn map_char(cur: &str) -> Option<&'static str> {
-    let val = match cur {
-        "ā" => "A",
-        "ī" => "I",
-        "ū" => "U",
-        "ṛ" => "f",
-        "ṝ" => "F",
-        "ḷ" => "x",
-        "ḹ" => "X",
-        "ai" => "E",
-        "au" => "O",
-        "ṃ" => "M",
-        "ḥ" => "H",
-        "ṅ" => "N",
-        "kh" => "K",
-        "gh" => "G",
-        "ch" => "C",
-        "jh" => "J",
-        "ñ" => "Y",
-        "ṭ" => "w",
-        "ṭh" => "W",
-        "ḍ" => "q",
-        "ḍh" => "Q",
-        "th" => "T",
-        "dh" => "D",
-        "ph" => "P",
-        "bh" => "B",
-        "ṇ" => "R",
-        "ś" => "S",
-        "ṣ" => "z",
-        "ḻ" => "L",
-        &_ => return None,
-    };
-    Some(val)
+
+#[derive(Debug, Serialize, Deserialize)]
+struct Scheme {
+    encodings: Vec<Encoding>,
+    implicit: String,
 }
 
-/// Hackily transliterate from IAST to SLP1.
-fn iast_to_slp1(input: &str) -> String {
-    let chars: Vec<char> = input.chars().collect();
-    let mut ret = String::new();
-    let mut i = 0;
-    while i < chars.len() {
-        let mut next: Option<&str> = None;
-        let mut offset = 0;
-
-        // Search for matches against our mapping. The longest IAST glyph has two characters,
-        // so search up to length 2. Start with 2 first so that we match greedily.
-        for j in [2, 1] {
-            let limit = cmp::min(i + j, chars.len());
-            let cur = String::from_iter(&chars[i..limit]);
-            offset = limit - i;
-
-            next = map_char(cur.as_str());
-            if let Some(_s) = next {
-                break;
-            }
-        }
+// Node represents the input mnemonic.
+// mnem: string associated with the node. ex: a, k or kh
+// tok_val: token associated with the mnemonic in a scheme
+// suffix: indicates whether the current mnem is a prefix to another mnem
+// tok_type: is it consonant or vowel. Depending on the scheme we may need
+//           pop or push a character.
+// implicit: this field may not be required - since info is already in
+//           the scheme.
+#[allow(dead_code)]
+#[derive(Clone, Debug, Default)]
+pub struct Node {
+    mnem: String, // Input mnemonic
+    #[allow(dead_code)]
+    assoc: String, // If a vowel - then the associated sign.
```

Comment:
`vowel_sign` ? This can be an `Option<String>`

Reply:
Ok.